store/tooling/auth.go

```diff
@@ -0,0 +1,253 @@
+}
+
+// parseSnapcraftLoginFile parses the content of snapcraft <v7 exported
+// login credentials files.
+func parseSnapcraftLoginFile(data []byte, what string) (*authData, bool, error) {
+	errPrefix := fmt.Sprintf("invalid snapcraft login %s", what)
+
+	cfg := goconfigparser.New()
+	// XXX this seems to almost always succeed
+	if err := cfg.ReadString(string(data)); err != nil {
+		likely := data[0] == '['
+		return nil, likely, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	sec := snapcraftLoginSection()
+	macaroon, err := cfg.Get(sec, "macaroon")
+	if err != nil {
+		return nil, true, fmt.Errorf("%s: %s", errPrefix, err)
+	}
+	unboundDischarge, err := cfg.Get(sec, "unbound_discharge")
+	if err != nil {
+		return nil, true, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	if macaroon == "" || unboundDischarge == "" {
+		return nil, true, fmt.Errorf("%s: empty fields", errPrefix)
+	}
+	return &authData{
+		Macaroon:   macaroon,
+		Discharges: []string{unboundDischarge},
+	}, false, nil
+}
+
+// parseAuthBase64JSON parses snapcraft v7+ base64-encoded auth credential data.
+func parseAuthBase64JSON(data []byte, what string) (*authData, bool, error) {
+	jsonData := make([]byte, base64.StdEncoding.DecodedLen(len(data)))
+	dataLen, err := base64.StdEncoding.Decode(jsonData, data)
+	if err != nil {
+		return nil, false, fmt.Errorf("cannot decode base64-encoded auth %s: %v", what, err)
+	}
+	var m map[string]interface{}
+	if err := json.Unmarshal(jsonData[:dataLen], &m); err != nil {
+		return nil, true, fmt.Errorf("cannot unmarshal base64-decoded auth %s: %v", what, err)
+	}
+	r, _ := m["r"].(string)
+	d, _ := m["d"].(string)
+	t, _ := m["t"].(string)
+	switch {
+	case t == "u1-macaroon":
+		v, _ := m["v"].(map[string]interface{})
+		r, _ = v["r"].(string)
+		d, _ = v["d"].(string)
+		if r == "" || d == "" {
+			break
+		}
+		fallthrough
+	case r != "" && d != "":
+		return &authData{
+			Macaroon:   r,
+			Discharges: []string{d},
+		}, false, nil
+	case t == "macaroon":
+		v, _ := m["v"].(string)
+		if v != "" {
+			return &authData{
+				Scheme: "Macaroon",
+				Value:  v,
+			}, false, nil
+		}
+	case t == "bearer":
+		v, _ := m["v"].(string)
+		if v != "" {
+			return &authData{
+				Scheme: "Bearer",
+				Value:  v,
+			}, false, nil
+		}
+	}
+	return nil, true, fmt.Errorf("cannot recognize unmarshalled base64-decoded auth %s: no known field combination set", what)
+}
+
+// UbuntuOneCreds can authorize requests using the implicitly carried
+// SSO/U1 user credentials.
+type UbuntuOneCreds struct {
+	User auth.UserState
+}
+
+// expected interfaces
+var _ store.Authorizer = (*UbuntuOneCreds)(nil)
+var _ store.RefreshingAuthorizer = (*UbuntuOneCreds)(nil)
+
+func (c *UbuntuOneCreds) Authorize(r *http.Request, _ store.DeviceAndAuthContext, user *auth.UserState, _ *store.AuthorizeOptions) error {
+	return store.UserAuthorizer{}.Authorize(r, nil, &c.User, nil)
+}
+
+func (c *UbuntuOneCreds) HasAuth(_ *auth.UserState) bool {
+	return true
```

Comment:
A comment would be good here. Don't we have to check that the user is not `nil` at least?